go/parquet/pqarrow/schema.go

```diff
@@ -948,7 +960,15 @@ func getNestedFactory(origin, inferred arrow.DataType) func(fieldList []arrow.Fi
 func applyOriginalStorageMetadata(origin arrow.Field, inferred *SchemaField) (modified bool, err error) {
 	nchildren := len(inferred.Children)
 	switch origin.Type.ID() {
-	case arrow.EXTENSION, arrow.SPARSE_UNION, arrow.DENSE_UNION:
+	case arrow.EXTENSION:
+		extType := arrow.GetExtensionType(origin.Type.(arrow.ExtensionType).ExtensionName())
+		if extType == nil {
+			err = xerrors.Errorf("arrow: extension type %q not registered", origin.Type.Name())
+			return
+		}
+		inferred.Field.Type = extType
```

Comment:
why `GetExtensionType` using the name when you already have the extension type here via `origin.Type` ?

If the extension type wasn't registered, then when the schema was deserialized you'd only have the storage type anyway and never get into this spot. In addition if the extension type is a parameterized type, then you're losing the metadata that was serialized into the `ExtensionMetadataKey`. 

I think the better solution here would probably be to do something like:

```go
extType := origin.Type.(arrow.ExtensionType)
modified, err = applyOriginalStorageMetadata(arrow.Field{
                 Type: extType.StorageType(), 
                 Metadata: arrow.NewMetadata(
                            []string{ipc.ExtensionTypeKeyName, ExtensionMetadataKeyName},
                            []string{extType.ExtensionName(), extType.Serialize()}),
           }, inferred)
if err != nil {
    return
}

inferred.Field.Type = extType
modified = true
```

Or something to that extent. That way it can properly handle extension types that have children / are defined in terms of a struct/list/etc. If the extension type wasn't registered, that's pretty much what would happen anyways since it would deserialize the schema as the storage type + the metadata.

What do you think? Might also be worthwhile to add a testing extension type that is parameterized and whose storage type is one of the nested types just to ensure it is handled correctly. You can find examples of parameteric types in the `arrow/internal/testing/types/extension_types.go` file.